Source/stores.h

```diff
@@ -12,96 +12,140 @@
-	SmithPremiumBuy,
 	Gossip,
-	StorytellerIdentifyShow,
-	Tavern,
-	Drunk,
-	Barmaid,
+	Invalid,
 };
 
-/** Currently active store */
-extern TalkID ActiveStore;
+enum class ItemLocation {
+	Inventory,
+	Belt,
+	Body
+};
 
-/** Current index into PlayerItemIndexes/PlayerItems */
-extern DVL_API_FOR_TEST int CurrentItemIndex;
-/** Map of inventory items being presented in the store */
-extern int8_t PlayerItemIndexes[48];
-/** Copies of the players items as presented in the store */
-extern DVL_API_FOR_TEST Item PlayerItems[48];
+struct StoreMenuOption {
+	TalkID action;
+	std::string text;
+};
 
-/** Items sold by Griswold */
-extern Item SmithItems[SMITH_ITEMS];
-/** Number of premium items for sale by Griswold */
-extern int PremiumItemCount;
-/** Base level of current premium items sold by Griswold */
-extern int PremiumItemLevel;
-/** Premium items sold by Griswold */
-extern Item PremiumItems[SMITH_PREMIUM_ITEMS];
+struct TownerLine {
+	const std::string menuHeader;
+	const StoreMenuOption *menuOptions;
+	size_t numOptions;
+};
 
-/** Items sold by Pepin */
-extern Item HealerItems[20];
+struct IndexedItem {
+	Item *itemPtr;         // Pointer to the original item
+	ItemLocation location; // Location in the player's inventory (Inventory, Belt, or Body)
+	int index;             // Index in the corresponding array
+};
 
-/** Items sold by Adria */
-extern Item WitchItems[WITCH_ITEMS];
+enum class ResourceType {
+	Life,
+	Mana,
+	Invalid,
+};
 
-/** Current level of the item sold by Wirt */
-extern int BoyItemLevel;
-/** Current item sold by Wirt */
-extern Item BoyItem;
+extern TalkID ActiveStore;                                    // Currently active store
+extern DVL_API_FOR_TEST std::vector<IndexedItem> playerItems; // Pointers to player items, coupled with necessary information
+
+class TownerStore {
+public:
+	TownerStore(std::string name, TalkID buyBasic, TalkID buy, TalkID sell, TalkID special, ResourceType resource)
+	    : name(name)
+	    , buyBasic(buyBasic)
+	    , buy(buy)
+	    , sell(sell)
+	    , special(special)
+	    , resourceType(resource)
+	{
+	}
+
+	std::string name;
+	std::vector<Item> basicItems; // Used for the blacksmith store that only displays non-magical items
+	std::vector<Item> items;
+	uint8_t itemLevel;
+
+	TalkID buyBasic;
+	TalkID buy;
+	TalkID sell;
+	TalkID special;
+	ResourceType resourceType; // Resource type to restore for stores that restore player's resources
+};
 
-void AddStoreHoldRepair(Item *itm, int8_t i);
+extern TownerStore Blacksmith;
+extern TownerStore Healer;
+extern TownerStore Witch;
+extern TownerStore Boy;
+extern TownerStore Storyteller;
+extern TownerStore Barmaid;
 
-/** Clears premium items sold by Griswold and Wirt. */
+/* Clears premium items sold by Griswold and Wirt. */
 void InitStores();
-
-/** Spawns items sold by vendors, including premium items sold by Griswold and Wirt. */
+/* Spawns items sold by vendors, including premium items sold by Griswold and Wirt. */

```

Comment:
Use short form doc block for single line function comments.
```suggestion
/** Spawns items sold by vendors, including premium items sold by Griswold and Wirt. */
```